### PR TITLE
feat: support diffusion models.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Fixes https://github.com/huggingface/optimum-quanto/issues/252

### CODE REVIEW

**Comment by sayakpaul** [comment]
@dacorvo would love to have some initial thoughts on how I have approached it.

File: optimum/quanto/models/transformers_models.py [hunk 1]
```diff
@@ -163,6 +126,8 @@ def from_pretrained(cls, model_name_or_path: Union[str, os.PathLike]):
             # Look for the index of a sharded checkpoint
             checkpoint_file = os.path.join(model_name_or_path, SAFE_WEIGHTS_INDEX_NAME)
             if os.path.exists(checkpoint_file):
+                from . import ShardedStateDict
```
Comment (dacorvo) [inline optimum/quanto/models/transformers_models.py:129]: Same here

File: test/models/test_quantized_model_for_pixart.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+from tempfile import TemporaryDirectory
+
+import pytest
+import torch
+
+from optimum.quanto import QModuleMixin, is_diffusers_available, qint4, qint8
+
+
+def quantized_model_for_pixart(qtype, exclude):
+    from diffusers import PixArtTransformer2DModel
```
Comment (dacorvo) [inline test/models/test_quantized_model_for_pixart.py:11]: Having these imports here is required to avoid import errors when diffusers is not installed.

**Review by dacorvo — CHANGES_REQUESTED** [review]
You need to fix your commits or the CI won't run the tests.
I ran them manually on my host, and I got errors.

File: test/models/test_quantized_model_for_pixart.py (hunk 2 above)
Comment (sayakpaul) [reply test/models/test_quantized_model_for_pixart.py:11]: Yeah.

File: optimum/quanto/models/__init__.py [hunk 3]
```diff
@@ -13,11 +13,55 @@
 # limitations under the License.
 
 import importlib
+import os
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply optimum/quanto/models/__init__.py:16]: Done.

File: optimum/quanto/models/diffusers_models.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                )
+
+            # Create an empty model
+            config = cls.base_class.load_config(model_name_or_path)
+            with init_empty_weights():
+                model = cls.base_class.from_config(config)
+            # Look for the index of a sharded checkpoint
+            checkpoint_file = os.path.join(model_name_or_path, SAFE_WEIGHTS_INDEX_NAME)
+            if os.path.exists(checkpoint_file):
+                from . import ShardedStateDict
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply optimum/quanto/models/diffusers_models.py]: Done.

**Comment by sayakpaul** [comment]
@dacorvo `shared_dict.py` related changes and commit issues have been fixed. 

About the failing tests:

```bash
FAILED test/models/test_quantized_model_for_pixart.py::test_quantized_model_for_pixart[without_proj_out-qint4] - RuntimeError: dimensionality of sizes (4) must match dimensionality of strides (2)
FAILED test/models/test_quantized_model_for_pixart.py::test_quantized_model_for_pixart[without_proj_out-qint8] - assert False
FAILED test/models/test_quantized_model_for_pixart.py::test_quantized_model_for_pixart[with_proj_out-qint4] - RuntimeError: dimensionality of sizes (4) must match dimensionality of strides (2)
FAILED test/models/test_quantized_model_for_pixart.py::test_quantized_model_for_pixart[with_proj_out-qint8] - assert False
```

Yes, I am not really sure what's going on for the assertion tests as the assertions for params (before we check output assertions) seem to be all passing. 

About 

```
RuntimeError: dimensionality of sizes (4) must match dimensionality of strides (2)
```

Not sure either as it stems from `requantize(model, state_dict=state_dict, quantization_map=qmap)`. So, seeking guidance.

**Comment by dacorvo** [comment]
I fixed both test failures here:
https://github.com/huggingface/optimum-quanto/tree/feat-diffusers-models
The first one is a bug, the second is only a workaround: some initialization code is not run when reloading the quangtized model.

**Comment by sayakpaul** [comment]
Okay this line is the culprint:
https://github.com/huggingface/optimum-quanto/blob/f61768cd4ea7c42b0e77a44d5a5b0f1a2ec8ea8b/optimum/quanto/quantize.py#L129

It zeros out the named buffer (`pos_embed`) in this case.

**Comment by dacorvo** [comment]
> Okay this line is the culprint:
> 
> https://github.com/huggingface/optimum-quanto/blob/f61768cd4ea7c42b0e77a44d5a5b0f1a2ec8ea8b/optimum/quanto/quantize.py#L129
> 
> It zeros out the named buffer (`pos_embed`) in this case.

Yes. I think moving to cpu is required, but I am not so sure emptying the params and buffers is actually required.
I did not write this code personally, it was contributed: what do you think ?

**Comment by sayakpaul** [comment]
Agreed!

**Comment by sayakpaul** [comment]
Perhaps this could be removed in a separate PR and I can rebase once merged?

**Comment by dacorvo** [comment]
Yes, let's do that. I will need to check the implications in terms of device memory.

**Comment by sayakpaul** [comment]
Sure I can wait till then

**Comment by sayakpaul** [comment]
I have pushed a temporary workaround in https://github.com/huggingface/optimum-quanto/pull/255/commits/bf961293af1407551914033b2080cf3ae2c589bf. I think for the purpose of this PR, that should work.

**Comment by sayakpaul** [comment]
Failing tests seem unrelated?

**Comment by sayakpaul** [comment]
I also wrote this mini script to use the quantized model in a full pipeline context:

```python
import torch
from diffusers import PixArtSigmaPipeline
from optimum.quanto import QuantizedPixArtTransformer2DModel
import gc

def flush():
    """Wipes off memory."""
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.reset_max_memory_allocated()
    torch.cuda.reset_peak_memory_stats()

def bytes_to_giga_bytes(bytes):
    return f"{(bytes / 1024 / 1024 / 1024):.3f}"


if __name__ == "__main__":
    flush()

    transformer = QuantizedPixArtTransformer2DModel.from_pretrained("/raid/.cache/huggingface/pixart-sigma-fp8")
    transformer.to(device="cuda", dtype=torch.float16)
    pipe = PixArtSigmaPipeline.from_pretrained(
        "PixArt-alpha/PixArt-Sigma-XL-2-1024-MS", 
        transformer=None,
        torch_dtype=torch.float16,
    ).to("cuda")
    pipe.transformer = transformer

    prompt = "A small cactus with a happy face in the Sahara desert."
    image = pipe(prompt).images[0]
    image.save("fused.png")

    mem_bytes = torch.cuda.max_memory_allocated()
    print(f"Memory: {bytes_to_giga_bytes(mem_bytes)} GBs.")
```

Without the FP8 model, Memory: 14.451 GBs.
With the FP8 model, Memory: 13.882 GBs.

Massive size reductions:

```bash
ls -lh /raid/.cache/huggingface/pixart-sigma-fp8

852 juil. 25 04:42 config.json
587M juil. 25 04:42 diffusion_pytorch_model.safetensors
33K juil. 25 04:42 quanto_qmap.json
```

The original checkpoint is [2.44 GB](https://huggingface.co/PixArt-alpha/PixArt-Sigma-XL-2-1024-MS/tree/main/transformer). 

"/raid/.cache/huggingface/pixart-sigma-fp8" was derived using:

```python
from diffusers import PixArtTransformer2DModel
from optimum.quanto import QuantizedPixArtTransformer2DModel, qfloat8

model = PixArtTransformer2DModel.from_pretrained("PixArt-alpha/PixArt-Sigma-XL-2-1024-MS", subfolder="transformer")
qmodel = QuantizedPixArtTransformer2DModel.quantize(model, weights=qfloat8)
qmodel.save_pretrained("/raid/.cache/huggingface/pixart-sigma-fp8")
```

File: optimum/quanto/models/diffusers_models.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                checkpoint_file = os.path.join(model_name_or_path, SAFETENSORS_WEIGHTS_NAME)
+                if not os.path.exists(checkpoint_file):
+                    raise ValueError(f"No safetensor weights found in {model_name_or_path}.")
+                # Get state_dict from model checkpoint
+                state_dict = load_state_dict(checkpoint_file)
+
+            # Requantize and load quantized weights from state_dict
+            requantize(model, state_dict=state_dict, quantization_map=qmap)
+            # Hack: re-assign the initial position embeddings as `requantize()` destroys it.
+            model.pos_embed.pos_embed = initial_pos_embed
```
Comment (sayakpaul) [inline optimum/quanto/models/diffusers_models.py:175]: I guess we don't need this now?

File: test/models/test_quantized_model_for_pixart.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    from optimum.quanto import QuantizedPixArtTransformer2DModel
+
+    exclude = "proj_out" if exclude_proj_out else None
+    quantized = quantized_model_for_pixart(qtype, exclude)
+    with TemporaryDirectory() as tmpdir:
+        quantized.save_pretrained(tmpdir)
+        requantized = QuantizedPixArtTransformer2DModel.from_pretrained(tmpdir)
+
+    # FIXME: position embeddings are not initialized in requantized model
+    # requantized.pos_embed.pos_embed = quantized.pos_embed.pos_embed
```
Comment (sayakpaul) [inline test/models/test_quantized_model_for_pixart.py:89]:
```suggestion
```

File: optimum/quanto/models/diffusers_models.py (hunk 5 above)
Comment (dacorvo) [reply optimum/quanto/models/diffusers_models.py:175]: No we don't.

**Comment by dacorvo** [comment]
I made the choice to enforce conventional commits because I don't squash commits to be able to bisect when tracking regressions. I know it is a bit tedious, but it is perfectly legit to rebase interactively, reword and force-push.

**Comment by dacorvo** [comment]
So here unfortunately "Apply suggestions from code review" is rejected, but you can reword it to "review: apply suggestions".

**Comment by sayakpaul** [comment]
A little too much for my taste, will give you that :P

**Comment by dacorvo** [comment]
> A little too much for my taste, will give you that :P

As a matter of fact, I really don't care about the format of the commit message, as long as it is meaningful. Since I cannot run an LLM to check the commit messages, and still wanted to reject things like 'wip', 'try', and the likes, I decided to enforce the conventional commits format, as a way to indicate to the submitter that he should try to make his commits readable.

The loophole is that you can still do: "chore: wip" 🙄

**Comment by sayakpaul** [comment]
Ouf, everything passes then.

File: optimum/quanto/models/diffusers_models.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        freeze(model)
+        return cls(model)
+
+    @classmethod
+    def from_pretrained(cls, model_name_or_path: Union[str, os.PathLike]):
+        if cls.base_class is None:
+            raise ValueError("The `base_class` attribute needs to be configured.")
+
+        if not is_accelerate_available():
+            raise ValueError("Reloading a quantized transformers model requires the accelerate library.")
```
Comment (dacorvo) [inline optimum/quanto/models/diffusers_models.py:123]:
```suggestion
            raise ValueError("Reloading a quantized diffusers model requires the accelerate library.")
```

**Review by dacorvo — APPROVED** [review]
Thank you so much for this pull-request ! I know it tested your patience because your code was basically correct right from the beginning but you discovered two nasty bugs.
A small nit, but otherwise excellent !
